wordpress-mobile/WordPress-iOS
# [Jetpack] Use an infinite scrolling view for Jetpack landing screen

- Related #19230
- May fix #19402

## Description

This updates the Jetpack animated scrolling view by:

1. Rendering the `JetpackLandingScreenView` a few times up front
2. Inserting these rendered views into a `UIStackView` contained in a `UIScrollView`
3. Scrolling the `UIScrollView` with a `InfiniteScrollerViewDelegate` that controls speed and direction, polled by `CADisplayLink`
4. Resetting the `UIScrollView` offset upon reaching certain vertical thresholds

| Before | After |
| - | - |
| <video src="https://user-images.githubusercontent.com/2092798/194401278-738c6875-5b50-4383-a563-c3bd8acd5e62.MP4" /> | <video src="https://user-images.githubusercontent.com/2092798/194401399-8883df3f-0a49-4c48-bb62-401684b35740.MP4" /> |

<details>
<summary>iPad Before</summary>

https://user-images.githubusercontent.com/2092798/194410392-c872a55b-239d-4013-92ee-a5405ec16469.mp4

</details>

<details>
<summary>iPad After</summary>

https://user-images.githubusercontent.com/2092798/194410460-bca0d934-d646-4854-b856-7bdf8dc55dcf.mp4

</details>

**Note:** Typography tweaks for the prompts can be pushed to subsequent PRs. There have been some regressions to appearance due to refactoring.

## Testing
- Enable the [new landing screen](https://github.com/wordpress-mobile/WordPress-iOS/blob/5978e9dd7cb1b3f167f743e363950e96c051f358/WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift#L116) feature flag

### Smooth animations
1. Clean install and load the Jetpack app (or log out of existing sessions to reach the landing screen)
2. Observe the scrolling prompts
3. **Expect:** The prompts to smoothly scroll with minimal animation hitches or jumps
4. Tap "Continue with WordPress.com"
5. Tap the "back" arrow at the top left
6. **Expect:** The prompts continue to smoothly scroll without delay
7. Tap "enter your existing site address"
8. Tap the "back" arrow at the top left
9. **Expect:** The prompts continue to smoothly scroll without delay
10. Rotate the device
11. **Expect:** No changes to the screen orientation - it should remain locked in portrait mode.
12. **Expect:** Content infinitely repeats with no gaps ([change to +1 to test opposite direction](https://github.com/wordpress-mobile/WordPress-iOS/pull/19410/files#diff-8a0c4de379fab1f48ba01fc7a7db49c7b5563e5bef98445edb15241ac607ab54R159))
13. **Expect:** When content repeats (e.g. a scrolling threshold has been reached and the offset resets, creating the infinite scrolling effect), colors do not change.

### VoiceOver doesn't announce prompts

See: https://github.com/wordpress-mobile/WordPress-iOS/pull/19381

1. Clean install and load the Jetpack app (or log out of existing sessions to reach the landing screen)
2. Observe the scrolling prompts
3. Enable VoiceOver and try to interact with the prompts
4. **Expect:** VoiceOver doesn't announce or interact with prompts

**Items to test:**
- iPhone / iPad
- a11y
   - Prompt sizes aren't affected by Dynamic Type
- Light to dark mode transition
   - May resolve #19402

## Regression Notes
1. Potential unintended areas of impact
- The Jetpack landing screen only

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- Manual tests above

3. What automated tests I added (or what prevented me from doing so)
I didn't add automated tests at this time. Logic to test the infinite scroll view should be straightforward and we should add it in a follow-up PR.

PR submission checklist:
